dex/path.go

```diff
@@ -5,62 +5,52 @@ package dex
 
 import (
 	"os"
-	"os/user"
 	"path/filepath"
-	"runtime"
 	"strings"
 )
 
-// CleanAndExpandPath expands environment variables and leading ~ in the
-// passed path, cleans the result, and returns it.
+// CleanAndExpandPath expands environment variables and leading ~ in the passed
+// path, cleans the result, and returns it.
 func CleanAndExpandPath(path string) string {
 	// Nothing to do when no path is given.
 	if path == "" {
 		return path
 	}
 
-	// NOTE: The os.ExpandEnv doesn't work with Windows cmd.exe-style
-	// %VARIABLE%, but the variables can still be expanded via POSIX-style
-	// $VARIABLE.
-	path = os.ExpandEnv(path)
-
-	if !strings.HasPrefix(path, "~") {
-		return filepath.Clean(path)
+	dirName := ""
+
+	// This supports Windows cmd.exe-style %VARIABLE%.
+	if strings.HasPrefix(path, "%") {
+		// Split path into %VARIABLE% and path
+		pathArray := strings.SplitAfterN(path, "/", 2)
+		dirEnv := strings.ToUpper(strings.Trim(pathArray[0], "%/"))
+		path = pathArray[1]
+		dirName = os.Getenv(dirEnv)
+		if dirName == "" {
+			// This does not support Windows XP and before as
+			// they didn't have a LOCALAPPDATA.
+			dirName = os.Getenv("LOCALAPPDATA")
+		}
+
+		return filepath.Join(dirName, path)
 	}
 
-	// Expand initial ~ to the current user's home directory, or ~otheruser
-	// to otheruser's home directory.  On Windows, both forward and backward
-	// slashes can be used.
-	path = path[1:]
-
-	var pathSeparators string
-	if runtime.GOOS == "windows" {
-		pathSeparators = string(os.PathSeparator) + "/"
-	} else {
-		pathSeparators = string(os.PathSeparator)
+	if !strings.HasPrefix(path, "~") {
+		// NOTE: The os.ExpandEnv doesn't work with Windows cmd.exe-style
+		// %VARIABLE%, but the variables can still be expanded via POSIX-style
+		// $VARIABLE.
+		path = os.ExpandEnv(path)
+		path, _ = filepath.Abs(path)
+		return path
 	}
 
-	userName := ""
-	if i := strings.IndexAny(path, pathSeparators); i != -1 {
-		userName = path[:i]
-		path = path[i:]
+	if strings.HasPrefix(path, "~") {
+		dirName, _ = os.UserHomeDir()
 	}
```

Comment:
There are two error ignoring places introduced here. If these methods are necessary I would prefer change the method to also return an error.

It looks like `filepath.Clean` is replaced with `filepath.Abs` and use of the `user` package replaced with `os.UserHomeDir`. Are these recommended? This original function was lifted from dcrd, which was probably lifted from bitcoind, so the whole thing is probably dated.

Also, the second `if` here is always true so not needed.